feat(multiple-prop): added the ability to select multiple values

souporserious opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

If you're new to contributing to open source projects, you might find this free
video course helpful: http://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->
**What**:
Related to https://github.com/paypal/react-kadabra/issues/61
<!-- Why are these changes necessary? -->
**Why**:
So we can build more complex menus that allow selecting multiple values.
<!-- How were these changes implemented? -->
**How**:
I followed how the browser handles `multiple` and `value`. When `multiple` is `true`, we handle value as an array. I switched the API from `selectedItem` -> `value` because I think this is a little more clear since you are technically selecting the value of your item[s]. This is also how other select menus like react-select handle this so it seemed like a safe change. I changed `value` -> `inputValue` which I think is more clear as well.
<!-- Have you done all of these things?  -->
**Checklist**:
<!-- add "N/A" to the end of each line that's irrelevant to your changes -->
<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->
- [x] Documentation
- [x] Tests
- [x] Ready to be merged <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->
- [ ] Added myself to contributors table <!-- this is optional, see the contributing guidelines for instructions -->

<!-- feel free to add additional comments -->
I was able to build a select menu, a multi-select, an autocomplete, and a simple dropdown menu with these changes. It's pretty cool how powerful, but simple it is! I _think_ this is closing in on a pretty final API after this. But please, let me know what you think 😇  if all looks well, I'll update the rest of the examples. It's a pretty minimal breaking change. Mainly `value` -> `inputValue`.

kentcdodds commented on src/autocomplete.js:
```diff
@@ -8,18 +8,18 @@ import {cbToCb, composeEventHandlers, debounce, scrollIntoView} from './utils'
   }
 
   static defaultProps = {
-    defaultSelectedItem: null,
```
Should this be `defaultValue: null`? Even though it doesn't actually make a difference, I like having defaults set for anything that's not required.

souporserious replied:
My bad. Done 👍

kentcdodds commented on src/autocomplete.js:
```diff
@@ -164,38 +179,43 @@ class Autocomplete extends Component {
+      clearSelection,
+    } = this
+    return {
+      // prop getters
```
Thank you!

kentcdodds commented on src/autocomplete.js:
```diff
@@ -164,38 +179,43 @@ class Autocomplete extends Component {
-      getRootProps,
+      clearSelection,
+    } = this
+    return {
```
Could we expose `getItemFromIndex` while we're at it? Ref: https://github.com/paypal/react-kadabra/issues/71

souporserious replied:
done ✅

kentcdodds commented on src/autocomplete.js:
```diff
@@ -288,15 +308,19 @@ class Autocomplete extends Component {
       'aria-expanded': isOpen,
       autoComplete: 'off',
-      value: (value === null ? this.getValue(selectedItem) : value) || '',
+      value: this.state.inputValue,
```
Let's destructure this above. I like to destructure stuff at the top of the closure so it's easy to tell what this function uses.

souporserious replied:
Oops! On it.

kentcdodds commented on src/autocomplete.js:
```diff
@@ -308,14 +332,12 @@ class Autocomplete extends Component {
-      this.openMenu(() => {
-        this.highlightIndex()
-      })
+      this.openMenu()
```
Why this change?

souporserious replied:
I moved that logic into `toggleMenu` so anytime the menu is opened it will handled highlighting.

kentcdodds commented on src/autocomplete.js:
```diff
@@ -338,11 +360,10 @@ class Autocomplete extends Component {
+    this.setState({
       isOpen: false,
       highlightedIndex: null,
-      value: this.getValue(selectedItem),
```
The `inputValue` should be reset to the `getValue` of the currently selected item right?

souporserious replied:
Oops, that was a mistake, putting it back 😁

souporserious replied:
Ohh that wasn't a mistake. I handle that in `selectItem` now. I actually need to handle `multiple` and `inputValue`.

kentcdodds commented on src/autocomplete.js:
```diff
@@ -371,10 +396,6 @@ class Autocomplete extends Component {
     this.toggleMenu(false, cb)
   }
 
-  getValue = item => {
```
Why is this removed?

souporserious replied:
I moved it closer to the input methods :)

souporserious commented:
I added some other things I found 😁 it looks like the recent highlighting changes got lost in the new API from today. I added that back, as well as the button API we had before, which turned out nice because we now have access to the same handlers as Input uses 🎉

souporserious commented:
Do you think `value` should be `selectedValue` so it's a little more explicit? 🤔

kentcdodds commented:
> Do you think value should be selectedValue so it's a little more explicit? 🤔

Yes 👍

kentcdodds commented on src/autocomplete.js:
```diff
@@ -257,6 +285,30 @@ class Autocomplete extends Component {
+
+    ArrowUp: this.keyDownHandlers.ArrowUp,
+
+    Enter(event) {
```
Actually, I think I prefer "Enter" on the button to select the highlighted item.

I'd love to hear what @marcysutton has to say on the subject of accessibility. Right now you can hit the spacebar on the button to toggle the menu and the up/down arrow keys to change the highlighted item, then the enter key to select an item. If we add this event handler, I feel like the Enter key has too much going on:

1. It can open the menu
2. It can close the menu
3. It can select the highlighted item.

I'd rather just have the enter key select the highlighted item if one is highlighted (like it is currently). [Here's the interaction](https://codesandbox.io/s/9rWEmzEv8) as it stands today: 

![a11y](https://user-images.githubusercontent.com/1500684/28931817-985f6030-7834-11e7-9185-2889af9cfe7c.gif)

souporserious replied:
Totally 👍 I did that wrong. I meant for space to handle that functionality :) that's seems to be how the default browser select works.

marcysutton replied:
Yep, space bar is the default for buttons.

souporserious commented:
Alright. I _think_ I have everything done. I was wondering in `onChange` whether we should pass the `selectedValue` there as well for convenience. That way you can see what value was just selected/deselected as well as look at the current state of selections.

kentcdodds commented:
> I was wondering in onChange whether we should pass the selectedValue there as well for convenience.

Yes. This sounds like a good idea. Though what if an item is removed in a multi-select scenario?

souporserious commented:
This is why I think we almost need to pass it now. That way you can compare what was just selected with the current values to determine that. I ended up with [something like this for selectly](https://github.com/souporserious/selectly/blob/master/src/Select.jsx#L146-L150) that worked well at our work. I don't _think_ we need to pass as much as I was passing though.

kentcdodds commented:
Got it so the API would be:

`onChange(currentValue, previousValue)`?

kentcdodds commented:
I kinda like what you have there actually. Maybe we should do an object...

```js
onChange({selectedValue, previousValue})
```

souporserious commented:
I like that. I think that's enough info to do what you need 🤔 maybe this will be good enough for now and we can revisit this part if we need to.

kentcdodds commented:
I wonder if we could also provide what kind of change it was: `onChange({selectedValue, previousValue, type: Autocomplete.changeTypes.removal})`

And in `changeTypes` we'd have `removal`, `addition`

Would that be useful? Should we concern ourselves with that?

tannermares commented:
@souporserious: I'm torn on wether or not the select should be that smart too. I have needed that value a couple times but wasn't super annoyed I had to do the subtraction myself to figure it out. I'm afraid it'll start a trend of random helper props that just bloats things. I love where this is going either way!

kentcdodds commented:
> I'm afraid it'll start a trend of random helper props that just bloats things

Yeah, let's skip the `type` thing and just stick with `selectedValue` and `previousValue`. We can always add it later. Harder to remove it ;-)

souporserious commented:
Cool :) that's what I was thinking. It's always easier to add than remove.

souporserious commented:
I landed on this. Thoughts?
![image](https://user-images.githubusercontent.com/2762082/28936747-80295ec2-783d-11e7-932b-cd27e52f0c76.png)

kentcdodds commented:
Cool. So `itemValue` would be the item that was just selected? What would this look like the context of a non-multiselect? What would `itemValue` be if an item is removed?

souporserious commented:
Haha I just deleted that comment 😆 I think I was doing pretty much what we talked about above. Let's see how `selectedValue` and `previousValue` play out.

souporserious commented:
Alright, I _think_ it's finally done 😁

kentcdodds commented on src/autocomplete.js:
```diff
@@ -137,11 +140,36 @@ class Autocomplete extends Component {
+          }
+          return {
+            selectedValue: values,
+            inputValue: values.map(value => this.getValue(value)).join(', '),
```
In the future I expect we'll need to make this a prop as well `getInputValue`... Either that or instead of mapping all the values to `this.getValue` we just call `this.getValue(values)`... Seems like that'd be reasonable... Should we do that now?

souporserious replied:
Ah yes, that's true. The end user might want to handle arrays in a different way or maybe they wouldn't want to, I'm not sure 🤔

souporserious replied:
Maybe ok to push through and see how people react? Keeps the API a little easier IMO

kentcdodds replied:
Sounds fine.

kentcdodds commented:
Sorry, I'm going to push forward with the rename. I want to get that done to avoid confusing people longer! Hopefully there will be no merge conflicts.

souporserious commented:
Looks like no merge conflicts 🙌

kentcdodds commented:
I'll merge this now. Release coming when I finish with my late lunch

kentcdodds commented:
```
~/Developer/downshift (master)
🐑  $ npm publish --tag beta
+ downshift@1.0.0-beta.9
```
